extlib/berkeleydb/src/read/index.rs

```diff
@@ -0,0 +1,93 @@
+use std::{io::Read, vec};
+
+use byteorder::{BigEndian, ByteOrder, LittleEndian, ReadBytesExt};
+use stable_eyre::{
+    eyre::{bail, Context},
+    Result,
+};
+
+use super::header::Header;
+
+pub struct Index(Vec<u16>);
+
+impl Index {
+    /// https://github.com/jssblck/go-rpmdb/blob/956701287363101ee9ade742d6bf1d5c5495f62a/pkg/bdb/constants.go#L10
+    const ENTRY_SIZE: usize = 2;
+
+    /// https://github.com/jssblck/go-rpmdb/blob/160242deff7a9ee82d1b493b62b7e50fd4c3e81c/pkg/bdb/hash_page.go#L102
+    const PAIR_SIZE: usize = Self::ENTRY_SIZE * 2;
+
+    /// Read [`Self`] out of a file in plain byte order.
+    ///
+    /// Reference: https://github.com/jssblck/go-rpmdb/blob/160242deff7a9ee82d1b493b62b7e50fd4c3e81c/pkg/bdb/hash_page.go#L89
+    pub fn parse<E: ByteOrder>(data: &[u8], entries: usize) -> Result<Self> {
+        if entries % 2 != 0 {
+            bail!("entries must only be requested in pairs: {entries}");
+        }
```

Comment:
[optional]: In general:
```rust
if $negative_condition {
  bail!($err, $args*);
}
```

is identical to `ensure!($positive_condition, $err, $args*);`

For example:
```suggestion
        ensure!(entries % 2 == 0, "entries must only be requested in pairs: {entries}");
```

This is identical, so it's not at all blocking, but I personally prefer reading `ensure!` over `if -> bail!`, so that I can gloss over it while reading data-flow logic, and read it when I care about validation logic.  I think one-liners are easier to ignore for readers who know what they're looking for.

There's quite a few of these in this PR, so I'm comfortable marking that as TODO or just leaving it.